core/src/main/java/org/infinispan/persistence/async/AsyncCacheWriter.java

```diff
@@ -242,14 +300,34 @@ public void run() {
                final boolean shouldStop;
                stateLock.readLock();
                try {
+                  if (!failSilently) {
+                     availabilityLock.lock();
+                     try {
+                        // If the delegate is unavailable, relinquish the readLock and await for the delegate to become available
+                        if (!delegateAvailable) {
+                           stateLock.readUnlock();
```

Comment:
I didn't want to keep the `stateLock` locked while we await the `availability` condition. But I have updated the code so that the `!failSilently` block now occurs before the stateLock's try/finally statement. That way the `stateLock` is only unlocked once when we continue the loop.